static/src/javascripts/projects/common/modules/video/ga-helper.ts

```diff
@@ -0,0 +1,105 @@
+type EventType =
+	| 'play'
+	| 'skip'
+	| '25'
+	| '50'
+	| '75'
+	| 'watched25'
+	| 'watched50'
+	| 'watched75'
+	| 'end';
+
+type Metrics =
+	| 'metric1'
+	| 'metric2'
+	| 'metric3'
+	| 'metric4'
+	| 'metric5'
+	| 'metric6';
+
+type EventMetrics = {
+	play?: 'metric1';
+	skip?: 'metric2';
+	'25'?: 'metric3';
+	'50'?: 'metric4';
+	'75'?: 'metric5';
+	watched25?: 'metric3';
+	watched50?: 'metric4';
+	watched75?: 'metric5';
+	end?: 'metric6';
+};
```

Comment:
If you want to keep this quite generic, you can use the [`Partial<T>`][] and [`Record<T>`][] utility types.


```suggestion
type EventMetrics = Partial<Record<EventType, Metrics>>
```

[`Partial<T>`]: https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype
[`Record<T>`]: https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type

Reply:
Alternatively, with a m
```suggestion
type EventMetrics = {
	[key in EventType]?: Metrics;
};
```

Reply:
Thanks for the suggestion! I don't think I knew you could do this.
From looking at the uses of `buildGoogleAnalyticsEvent` it seems like the `EventMetrics` s always conform to a pretty similar structure; `play` is always `metric1`, `skip` is `metric2`, etc. 
It seems to me that we might as well have our a types describe this structure as precisely as possible and disallow objects that don't match e.g. {'play': 'metric2'}. The downside is that we don't get to take advantage of `EventType` and `Metric` if we define `EventMetrics` from scratch 🙁

Reply:
Thinking about this a little more, you could do something like this ([TypeScript Playground][]):

```ts
const eventMetrics = {
	play: 'metric1',
	skip: 'metric2',
	'25': 'metric3',
	'50': 'metric4',
	'75': 'metric5',
	watched25: 'metric3',
	watched50: 'metric4',
	watched75: 'metric5',
	end: 'metric6',
} as const;

type EventKey = keyof typeof eventMetrics;
type Metric = typeof eventMetrics[EventKey];
```

[TypeScript Playground]: https://www.typescriptlang.org/play?#code/MYewdgzgLgBApgNzmKBZOUBOBLYEYC8MA3gFACQADgDYCGAngFwwDkAthjsAIwsA0FCAGtslZu064ATPwospAVhbiOWXAGZZ5FgoAMy1qq4AWLSwDsSlZOBKB5AO60owABZwAJoutrgm+04u7h56PiZagW6elmG4dhTIHrHAAGyyAL4wtPigkFAA3KSkUPSUcDAAokgoANJw9IQwQvUgAGYwJWVt8NVoNhCFneXovo1D3YjIfb4QANpVU3X0ALr5QA

Reply:
Yes, this is much better! Thanks for the suggestion
Made this change in #24386